Sources/MapboxCoreNavigation/MapboxRoutingProvider.swift

```diff
@@ -0,0 +1,338 @@
+        
+        guard let responseIdentifier = indexedRouteResponse.routeResponse.identifier else {
+            defer {
+                completionHandler(session, .failure(.noData))
+            }
+            return nil
+        }
+        
+        let encoder = JSONEncoder()
+        encoder.userInfo[.options] = routeOptions
+        
+        let routeIndex = UInt32(indexedRouteResponse.routeIndex)
+        
+        guard let routeData = try? encoder.encode(indexedRouteResponse.routeResponse),
+              let routeJSONString = String(data: routeData, encoding: .utf8) else {
+            preconditionFailure("Could not serialize route data for refreshing.")
+        }
+        
+        var requestId: RequestId!
+        let refreshOptions = RouteRefreshOptions(requestId: responseIdentifier,
+                                                 routeIndex: routeIndex,
+                                                 legIndex: startLegIndex,
+                                                 routingProfile: routeOptions.profileIdentifier.nativeProfile)
+        requestsLock.lock()
+        requestId = router.getRouteRefresh(for: refreshOptions,
+                                           route: routeJSONString) { [weak self] result, _ in
+            guard let self = self else { return }
+            
+            self.parseResponse(requestId: requestId,
+                               userInfo: [.options: routeOptions,
+                                          .credentials: self.settings.directions.credentials],
+                               result: result) { (response: Result<RouteResponse, DirectionsError>) in
+                completionHandler(session, response)
+            }
+        }
+        activeRequests[requestId] = Request(requestIdentifier: requestId,
+                                                   router: self)
+        requestsLock.unlock()
+        return activeRequests[requestId]
+    }
+    
+    // MARK: - Private methods
+    
+    fileprivate func complete(requestId: RequestId, with result: @escaping () -> Void) {
+        DispatchQueue.main.async { [self] in
+            result()
+            
+            requestsLock.lock(); defer {
+                requestsLock.unlock()
+            }
+            
+            activeRequests[requestId] = nil
+        }
+    }
+
+    struct ResponseDisposition: Decodable {
+        var code: String?
+        var message: String?
+        var error: String?
+        
+        private enum CodingKeys: CodingKey {
+            case code, message, error
+        }
+    }
+    
+    fileprivate func parseResponse<ResponseType: Codable>(requestId: RequestId, userInfo: [CodingUserInfoKey : Any], result: Expected<AnyObject, AnyObject>, completion: @escaping (Result<ResponseType, DirectionsError>) -> Void) {
+        do {
+            let json = result.value as? String
+            guard let data = json?.data(using: .utf8) else {
```

Comment:
Sorry, I didn't get your suggestion

Reply:
```swift
guard let json = result.value as? String else {...}
let data = Data(json.utf8)
...
```

Reply:
Why is it preferable?

Reply:
`Data(json.utf8)` isn't optional. 
`data(using: .utf8)` is optional. 

It is more clear that this code can fail because result value is of incorrect type. In your variant it is hidden, but the one who knows that data method returns nullable, it is not. 

Just noticed, that you could use `Result+Expected.swift` for parsing NavNative's Expected type.